src/protocols/secure_channel/CASESession.cpp

```diff
@@ -638,6 +655,45 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
     return CHIP_NO_ERROR;
 }
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+void CASESession::HandleConnectionComplete(void * context, Transport::ActiveConnectionState * conObj, CHIP_ERROR conErr)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrReturn(context != nullptr);
+    VerifyOrReturn(conObj != nullptr);
+    VerifyOrReturn(conObj->mPeerAddr.GetTransportType() == Transport::Type::kTcp);
+
+    ChipLogDetail(SecureChannel, "TCP Connection established before session establishment");
+
+    CASESession * caseSession = reinterpret_cast<CASESession *>(context);
+
+    // Associate the connection with the current unauthenticated session for the
+    // CASE exchange.
+    caseSession->mExchangeCtxt->GetSessionHandle()->AsUnauthenticatedSession()->SetTCPConnection(conObj);
+
+    // Associate the connection with the current secure session that is being
+    // set up.
+    caseSession->mSecureSessionHolder.Get().Value()->AsSecureSession()->SetTCPConnection(conObj);
+
+    // Retain the initiator exchange until CASE interaction is complete.
+    // Exchange released in PairingSession::Finish()
+    caseSession->mExchangeCtxt->Retain();
+
+    // Send Sigma1 after connection is established for sessions over TCP
+    err = caseSession->SendSigma1();
+    if (err != CHIP_NO_ERROR)
+    {
+        caseSession->Clear();
```

Comment:
What happens to the extra retain we did on the exchange above?  Do we just leak the exchange?  Or does the error case here end up calling Finish()?

But also: why do we need to do the extra retain on the exchange at all?  This does not make sense to me.